File: src/contexten/backend/routes/github.js
const express = require('express');
const axios = require('axios');
const router = express.Router();

// GitHub OAuth credentials
const GITHUB_CLIENT_ID = process.env.GITHUB_CLIENT_ID;
const GITHUB_CLIENT_SECRET = process.env.GITHUB_CLIENT_SECRET;
const GITHUB_REDIRECT_URI = process.env.GITHUB_REDIRECT_URI || 'http://localhost:3000/github/callback';

/**
 * Exchange GitHub OAuth code for access token
 * POST /api/github/callback
 */
router.post('/callback', async (req, res) => {
  try {
    const { code } = req.body;
    
    if (!code) {
      return res.status(400).json({ 
        success: false, 
        message: 'Authorization code is required' 
      });
    }
    
    // Exchange code for access token
    const tokenResponse = await axios.post(
      'https://github.com/login/oauth/access_token',
      {
        client_id: GITHUB_CLIENT_ID,
        client_secret: GITHUB_CLIENT_SECRET,
        code,
        redirect_uri: GITHUB_REDIRECT_URI,
      },
      {
        headers: {
          Accept: 'application/json',
        },
      }
    );
    
    const { access_token, token_type, scope } = tokenResponse.data;
    
    if (!access_token) {
      return res.status(400).json({ 
        success: false, 
        message: 'Failed to obtain access token' 
      });
    }
    
    // Get user information
    const userResponse = await axios.get('https://api.github.com/user', {
      headers: {
        Authorization: `token ${access_token}`,
      },
    });
    
    return res.json({
      success: true,
      message: 'GitHub authentication successful',
      data: {
        token: access_token,
        token_type,
        scope,
        user: userResponse.data,
      },
    });
  } catch (error) {
    console.error('GitHub callback error:', error);
    return res.status(500).json({
      success: false,
      message: 'Error processing GitHub authentication',
      error: error.message,
    });
  }
});

/**
 * Get repositories for the authenticated user
 * GET /api/github/repositories
 */
router.get('/repositories', async (req, res) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];
    
    if (!token) {
      return res.status(401).json({ 
        success: false, 
        message: 'GitHub token is required' 
      });
    }
    
    const response = await axios.get('https://api.github.com/user/repos', {
      headers: {
        Authorization: `token ${token}`,
      },
      params: {
        sort: 'updated',
        per_page: 100,
      },
    });
    
    return res.json({
      success: true,
      message: 'Repositories retrieved successfully',
      data: {
        repositories: response.data,
      },
    });
  } catch (error) {
    console.error('Error fetching GitHub repositories:', error);
    return res.status(500).json({
      success: false,
      message: 'Error fetching GitHub repositories',
      error: error.message,
    });
  }
});

/**
 * Get user information
 * GET /api/github/user
 */
router.get('/user', async (req, res) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];
    
    if (!token) {
      return res.status(401).json({ 
        success: false, 
        message: 'GitHub token is required' 
      });
    }
    
    const response = await axios.get('https://api.github.com/user', {
      headers: {
        Authorization: `token ${token}`,
      },
    });
    
    return res.json({
      success: true,
      message: 'User information retrieved successfully',
      data: {
        user: response.data,
      },
    });
  } catch (error) {
    console.error('Error fetching GitHub user:', error);
    return res.status(500).json({
      success: false,
      message: 'Error fetching GitHub user information',
      error: error.message,
    });
  }
});

module.exports = router;

File: src/app/auth/github.py
"""
GitHub OAuth authentication handler.
"""

from typing import Optional, Dict, Any
import aiohttp
from fastapi import HTTPException, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Session

from ..core.config import settings
from ..db.base import get_db
from ..db.models import User

class GitHubOAuth:
    """GitHub OAuth handler."""

    def __init__(self):
        self.client_id = settings.GITHUB_CLIENT_ID
        self.client_secret = settings.GITHUB_CLIENT_SECRET
        self.callback_url = settings.GITHUB_CALLBACK_URL

    async def get_access_token(self, code: str) -> str:
        """Exchange code for access token."""
        async with aiohttp.ClientSession() as session:
            async with session.post(
                "https://github.com/login/oauth/access_token",
                json={
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                    "code": code,
                    "redirect_uri": self.callback_url,
                },
                headers={"Accept": "application/json"},
            ) as response:
                if response.status != 200:
                    raise HTTPException(status_code=400, detail="Failed to get access token")
                data = await response.json()
                return data.get("access_token")

    async def get_user_info(self, access_token: str) -> Dict[str, Any]:
        """Get GitHub user information."""
        async with aiohttp.ClientSession() as session:
            async with session.get(
                "https://api.github.com/user",
                headers={
                    "Authorization": f"token {access_token}",
                    "Accept": "application/json",
                },
            ) as response:
                if response.status != 200:
                    raise HTTPException(status_code=400, detail="Failed to get user info")
                return await response.json()

    async def get_or_create_user(
        self, user_info: Dict[str, Any], db: AsyncSession
    ) -> User:
        """Get existing user or create new one."""
        # Check if user exists
        user = await db.query(User).filter(User.github_id == str(user_info["id"])).first()
        
        if not user:
            # Create new user
            user = User(
                github_id=str(user_info["id"]),
                email=user_info.get("email"),
                username=user_info["login"],
            )
            db.add(user)
            await db.commit()
            await db.refresh(user)
        
        return user

github_oauth = GitHubOAuth()

async def get_current_user(
    db: AsyncSession = Depends(get_db),
    token: str = Depends(get_access_token),
) -> Optional[User]:
    """Get current authenticated user."""
    if not token:
        return None
        
    try:
        user_info = await github_oauth.get_user_info(token)
        return await github_oauth.get_or_create_user(user_info, db)
    except Exception:
        return None

